lib/wallet/src/Cardano/Wallet/Write/Tx.hs

```diff
@@ -0,0 +1,481 @@
+    ( StandardCrypto )
+import Cardano.Ledger.Era
+    ( Crypto )
+import Cardano.Ledger.Mary
+    ( Value )
+import Cardano.Ledger.SafeHash
+    ( SafeHash, extractHash, unsafeMakeSafeHash )
+import Data.ByteString
+    ( ByteString )
+import Data.ByteString.Lazy
+    ( fromStrict )
+import Data.ByteString.Short
+    ( toShort )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Maybe.Strict
+    ( StrictMaybe (..) )
+
+import qualified Cardano.Api as Cardano
+import qualified Cardano.Api.Extra as Cardano
+import qualified Cardano.Api.Shelley as Cardano
+import qualified Cardano.Binary as CBOR
+import qualified Cardano.Crypto.Hash.Class as Crypto
+import qualified Cardano.Ledger.Address as Ledger
+import qualified Cardano.Ledger.Alonzo.Data as Alonzo
+import qualified Cardano.Ledger.Alonzo.Scripts as Alonzo
+import qualified Cardano.Ledger.Alonzo.TxBody as Alonzo
+import qualified Cardano.Ledger.Babbage as Babbage
+import qualified Cardano.Ledger.Babbage.TxBody as Babbage
+import qualified Cardano.Ledger.Core as Core
+import qualified Cardano.Ledger.Shelley.UTxO as Shelley
+import qualified Cardano.Ledger.TxIn as Ledger
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Types as Aeson
+import qualified Data.Map as Map
+
+--------------------------------------------------------------------------------
+-- Eras
+--------------------------------------------------------------------------------
+
+type LatestEra = BabbageEra
+
+type LatestLedgerEra = Cardano.ShelleyLedgerEra LatestEra
+
+--------------------------------------------------------------------------------
+-- RecentEra
+--------------------------------------------------------------------------------
+
+-- NOTE: We /could/ let 'era' refer to eras from the ledger rather than from
+-- cardano-api.
+data RecentEra era where
+    RecentEraBabbage :: RecentEra BabbageEra
+    RecentEraAlonzo :: RecentEra AlonzoEra
+
+class Cardano.IsShelleyBasedEra era => IsRecentEra era where
+    recentEra :: RecentEra era
+
+-- | Return a proof that the wallet can create txs in this era, or @Nothing@.
+isRecentEra :: Cardano.CardanoEra era -> Maybe (RecentEra era)
+isRecentEra Cardano.BabbageEra = Just RecentEraBabbage
+isRecentEra Cardano.AlonzoEra = Just RecentEraAlonzo
+isRecentEra Cardano.MaryEra = Nothing
+isRecentEra Cardano.AllegraEra = Nothing
+isRecentEra Cardano.ShelleyEra = Nothing
+isRecentEra Cardano.ByronEra = Nothing
+
+instance IsRecentEra BabbageEra where
+    recentEra = RecentEraBabbage
+
+instance IsRecentEra AlonzoEra where
+    recentEra = RecentEraAlonzo
+
+cardanoEraFromRecentEra :: RecentEra era -> Cardano.CardanoEra era
+cardanoEraFromRecentEra =
+    Cardano.shelleyBasedToCardanoEra
+    . shelleyBasedEraFromRecentEra
+
+shelleyBasedEraFromRecentEra :: RecentEra era -> Cardano.ShelleyBasedEra era
+shelleyBasedEraFromRecentEra RecentEraBabbage = Cardano.ShelleyBasedEraBabbage
+shelleyBasedEraFromRecentEra RecentEraAlonzo = Cardano.ShelleyBasedEraAlonzo
```

Comment:
Suggestion: use `LambdaCase` for shorter code.

Reply:
I'm a bit torn in this case — I think the above way reads simpler in a way — but sure, going with it.